File: src/services/reconciliationService.ts
import { apiService } from './apiService';

export interface ReconciliationTask {
  id: string;
  ticket_id: string;
  client_id: string;
  client_name: string;
  corrections: CorrectionItem[];
  status: 'pending' | 'reconciled' | 'failed' | 'partial';
  created_at: string;
  reconciled_at?: string;
  attempts: number;
  last_attempt_at?: string;
  error_message?: string;
}

export interface CorrectionItem {
  field: string;
  field_label: string;
  expected_value: string;
  cbs_value: string | null;
  is_matched: boolean;
  last_checked_at?: string;
}

export interface ReconciliationResult {
  task_id: string;
  status: 'success' | 'partial' | 'failed';
  matched_fields: number;
  total_fields: number;
  discrepancies: Discrepancy[];
  checked_at: string;
}

export interface Discrepancy {
  field: string;
  field_label: string;
  expected_value: string;
  actual_value: string | null;
  severity: 'low' | 'medium' | 'high';
}

export interface ReconciliationStats {
  total_pending: number;
  reconciled_today: number;
  failed_today: number;
  success_rate: number;
  average_reconciliation_time: number;
  by_status: {
    status: string;
    count: number;
  }[];
}

class ReconciliationService {
  async getPendingReconciliations(filters?: {
    agency_code?: string;
    client_id?: string;
  }): Promise<ReconciliationTask[]> {
    try {
      const params = new URLSearchParams();
      if (filters?.agency_code) params.append('agency_code', filters.agency_code);
      if (filters?.client_id) params.append('client_id', filters.client_id);

      const response = await apiService.get<ReconciliationTask[]>(
        `/api/reconciliation/pending?${params.toString()}`
      );
      return response;
    } catch (error) {
      console.error('Error fetching pending reconciliations:', error);
      return [];
    }
  }

  async getReconciliationById(id: string): Promise<ReconciliationTask | null> {
    try {
      return await apiService.get<ReconciliationTask>(`/api/reconciliation/${id}`);
    } catch (error) {
      console.error('Error fetching reconciliation:', error);
      return null;
    }
  }

  async reconcileTask(taskId: string): Promise<ReconciliationResult | null> {
    try {
      return await apiService.post<ReconciliationResult>(
        `/api/reconciliation/${taskId}/reconcile`,
        {}
      );
    } catch (error) {
      console.error('Error reconciling task:', error);
      return null;
    }
  }

  async reconcileAll(filters?: {
    agency_code?: string;
    max_tasks?: number;
  }): Promise<{ success: number; failed: number; total: number }> {
    try {
      const response = await apiService.post<{ success: number; failed: number; total: number }>(
        '/api/reconciliation/reconcile-all',
        filters || {}
      );
      return response;
    } catch (error) {
      console.error('Error reconciling all tasks:', error);
      return { success: 0, failed: 0, total: 0 };
    }
  }

  async retryReconciliation(taskId: string): Promise<ReconciliationResult | null> {
    try {
      return await apiService.post<ReconciliationResult>(
        `/api/reconciliation/${taskId}/retry`,
        {}
      );
    } catch (error) {
      console.error('Error retrying reconciliation:', error);
      return null;
    }
  }

  async closeTicket(ticketId: string, userId: string, comments: string): Promise<boolean> {
    try {
      await apiService.post(`/api/reconciliation/${ticketId}/close`, {
        user_id: userId,
        comments,
      });
      return true;
    } catch (error) {
      console.error('Error closing ticket:', error);
      return false;
    }
  }

  async getReconciliationStats(agencyCode?: string): Promise<ReconciliationStats | null> {
    try {
      const url = agencyCode
        ? `/api/reconciliation/stats?agency_code=${agencyCode}`
        : '/api/reconciliation/stats';
      return await apiService.get<ReconciliationStats>(url);
    } catch (error) {
      console.error('Error fetching reconciliation stats:', error);
      return null;
    }
  }

  async getReconciliationHistory(filters?: {
    ticket_id?: string;
    client_id?: string;
    start_date?: string;
    end_date?: string;
    status?: string;
  }): Promise<ReconciliationTask[]> {
    try {
      const params = new URLSearchParams();
      if (filters?.ticket_id) params.append('ticket_id', filters.ticket_id);
      if (filters?.client_id) params.append('client_id', filters.client_id);
      if (filters?.start_date) params.append('start_date', filters.start_date);
      if (filters?.end_date) params.append('end_date', filters.end_date);
      if (filters?.status) params.append('status', filters.status);

      const response = await apiService.get<ReconciliationTask[]>(
        `/api/reconciliation/history?${params.toString()}`
      );
      return response;
    } catch (error) {
      console.error('Error fetching reconciliation history:', error);
      return [];
    }
  }

  getStatusColor(status: string): string {
    switch (status) {
      case 'reconciled':
        return 'text-green-600 bg-green-100';
      case 'pending':
        return 'text-yellow-600 bg-yellow-100';
      case 'failed':
        return 'text-red-600 bg-red-100';
      case 'partial':
        return 'text-orange-600 bg-orange-100';
      default:
        return 'text-gray-600 bg-gray-100';
    }
  }

  getStatusLabel(status: string): string {
    const labels: Record<string, string> = {
      pending: 'En attente',
      reconciled: 'Réconcilié',
      failed: 'Échec',
      partial: 'Partiel',
    };
    return labels[status] || status;
  }

  calculateMatchPercentage(matched: number, total: number): number {
    if (total === 0) return 0;
    return Math.round((matched / total) * 100);
  }

  getSeverityColor(severity: string): string {
    switch (severity) {
      case 'high':
        return 'text-red-600 bg-red-100';
      case 'medium':
        return 'text-orange-600 bg-orange-100';
      case 'low':
        return 'text-yellow-600 bg-yellow-100';
      default:
        return 'text-gray-600 bg-gray-100';
    }
  }

  getSeverityLabel(severity: string): string {
    const labels: Record<string, string> = {
      high: 'Haute',
      medium: 'Moyenne',
      low: 'Basse',
    };
    return labels[severity] || severity;
  }
}

export const reconciliationService = new ReconciliationService();

File: backend-java/src/main/java/com/bsic/dataqualitybackend/service/ReconciliationService.java
package com.bsic.dataqualitybackend.service;

import com.bsic.dataqualitybackend.repository.InformixRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Slf4j
@ConditionalOnProperty(name = "app.features.informix-integration", havingValue = "true", matchIfMissing = false)
public class ReconciliationService {

    private final InformixRepository informixRepository;

    @Qualifier("primaryJdbcTemplate")
    private final JdbcTemplate mysqlJdbcTemplate;

    public Map<String, Object> reconcileTask(String taskId) {
        log.info("Starting reconciliation for task: {}", taskId);

        Map<String, Object> task = getTask(taskId);
        if (task == null) {
            throw new RuntimeException("Task not found: " + taskId);
        }

        String clientId = (String) task.get("client_id");
        String ticketId = (String) task.get("ticket_id");

        Map<String, Object> cbsData = informixRepository.getClientById(clientId);
        if (cbsData == null || cbsData.isEmpty()) {
            throw new RuntimeException("Client not found in CBS: " + clientId);
        }

        List<Map<String, Object>> corrections = getCorrections(ticketId);
        List<Map<String, Object>> discrepancies = new ArrayList<>();
        int matchedFields = 0;

        for (Map<String, Object> correction : corrections) {
            String field = (String) correction.get("field_name");
            String expectedValue = (String) correction.get("new_value");
            String cbsColumn = mapFieldToCBSColumn(field);
            Object cbsValueObj = cbsData.get(cbsColumn);
            String cbsValue = cbsValueObj != null ? cbsValueObj.toString().trim() : "";

            boolean isMatched = compareValues(expectedValue, cbsValue);

            if (isMatched) {
                matchedFields++;
            } else {
                Map<String, Object> discrepancy = new HashMap<>();
                discrepancy.put("field", field);
                discrepancy.put("field_label", correction.get("field_label"));
                discrepancy.put("expected_value", expectedValue);
                discrepancy.put("actual_value", cbsValue);
                discrepancy.put("severity", calculateSeverity(field));
                discrepancies.add(discrepancy);
            }

            updateCorrectionStatus(ticketId, field, cbsValue, isMatched);
        }

        int totalFields = corrections.size();
        String status = matchedFields == totalFields ? "reconciled"
                : matchedFields > 0 ? "partial"
                : "failed";

        updateTaskStatus(taskId, status);

        Map<String, Object> result = new HashMap<>();
        result.put("task_id", taskId);
        result.put("status", "reconciled".equals(status) ? "success" : status);
        result.put("matched_fields", matchedFields);
        result.put("total_fields", totalFields);
        result.put("discrepancies", discrepancies);
        result.put("checked_at", LocalDateTime.now().toString());

        log.info("Reconciliation completed for task {}: {}/{} fields matched",
                taskId, matchedFields, totalFields);

        return result;
    }

    public List<Map<String, Object>> getPendingTasks(String agencyCode, String clientId) {
        StringBuilder sql = new StringBuilder("""
            SELECT DISTINCT
                t.id, t.ticket_id, t.client_id, t.status, t.created_at,
                t.attempts, t.last_attempt_at, t.error_message,
                a.client_name, a.agency_code
            FROM reconciliation_tasks t
            LEFT JOIN tickets tk ON tk.ticket_number = t.ticket_id
            LEFT JOIN anomalies a ON a.id = tk.anomaly_id
            WHERE t.status = 'pending'
        """);

        List<Object> params = new ArrayList<>();

        if (agencyCode != null && !agencyCode.isEmpty()) {
            sql.append(" AND a.agency_code = ?");
            params.add(agencyCode);
        }

        if (clientId != null && !clientId.isEmpty()) {
            sql.append(" AND t.client_id = ?");
            params.add(clientId);
        }

        sql.append(" ORDER BY t.created_at DESC LIMIT 100");

        List<Map<String, Object>> tasks = mysqlJdbcTemplate.queryForList(
                sql.toString(),
                params.toArray()
        );

        for (Map<String, Object> task : tasks) {
            String ticketId = (String) task.get("ticket_id");
            List<Map<String, Object>> corrections = getCorrections(ticketId);
            task.put("corrections", corrections);
        }

        return tasks;
    }

    public List<Map<String, Object>> getReconciliationHistory(
            String ticketId, String clientId, String status,
            LocalDateTime startDate, LocalDateTime endDate) {

        StringBuilder sql = new StringBuilder("""
            SELECT DISTINCT
                t.id, t.ticket_id, t.client_id, t.status, t.created_at,
                t.reconciled_at, t.attempts, t.last_attempt_at, t.error_message,
                a.client_name, a.agency_code
            FROM reconciliation_tasks t
            LEFT JOIN tickets tk ON tk.ticket_number = t.ticket_id
            LEFT JOIN anomalies a ON a.id = tk.anomaly_id
            WHERE 1=1
        """);

        List<Object> params = new ArrayList<>();

        if (ticketId != null && !ticketId.isEmpty()) {
            sql.append(" AND t.ticket_id = ?");
            params.add(ticketId);
        }

        if (clientId != null && !clientId.isEmpty()) {
            sql.append(" AND t.client_id = ?");
            params.add(clientId);
        }

        if (status != null && !status.isEmpty()) {
            sql.append(" AND t.status = ?");
            params.add(status);
        }

        if (startDate != null) {
            sql.append(" AND t.created_at >= ?");
            params.add(startDate);
        }

        if (endDate != null) {
            sql.append(" AND t.created_at <= ?");
            params.add(endDate);
        }

        sql.append(" ORDER BY t.created_at DESC LIMIT 500");

        List<Map<String, Object>> tasks = mysqlJdbcTemplate.queryForList(
                sql.toString(),
                params.toArray()
        );

        for (Map<String, Object> task : tasks) {
            String tktId = (String) task.get("ticket_id");
            List<Map<String, Object>> corrections = getCorrections(tktId);
            task.put("corrections", corrections);
        }

        return tasks;
    }

    public Map<String, Object> getStats(String agencyCode) {
        String whereClause = agencyCode != null && !agencyCode.isEmpty()
                ? "WHERE a.agency_code = ?"
                : "WHERE 1=1";

        String sql = String.format("""
            SELECT
                COUNT(CASE WHEN t.status = 'pending' THEN 1 END) as total_pending,
                COUNT(CASE WHEN t.status = 'reconciled' AND DATE(t.reconciled_at) = CURDATE() THEN 1 END) as reconciled_today,
                COUNT(CASE WHEN t.status = 'failed' AND DATE(t.last_attempt_at) = CURDATE() THEN 1 END) as failed_today,
                ROUND(
                    COUNT(CASE WHEN t.status = 'reconciled' THEN 1 END) * 100.0 /
                    NULLIF(COUNT(*), 0), 2
                ) as success_rate,
                AVG(TIMESTAMPDIFF(SECOND, t.created_at, t.reconciled_at)) as average_reconciliation_time
            FROM reconciliation_tasks t
            LEFT JOIN tickets tk ON tk.ticket_number = t.ticket_id
            LEFT JOIN anomalies a ON a.id = tk.anomaly_id
            %s
        """, whereClause);

        Map<String, Object> stats = agencyCode != null && !agencyCode.isEmpty()
                ? mysqlJdbcTemplate.queryForMap(sql, agencyCode)
                : mysqlJdbcTemplate.queryForMap(sql);

        String statusSql = String.format("""
            SELECT t.status, COUNT(*) as count
            FROM reconciliation_tasks t
            LEFT JOIN tickets tk ON tk.ticket_number = t.ticket_id
            LEFT JOIN anomalies a ON a.id = tk.anomaly_id
            %s
            GROUP BY t.status
        """, whereClause);

        List<Map<String, Object>> byStatus = agencyCode != null && !agencyCode.isEmpty()
                ? mysqlJdbcTemplate.queryForList(statusSql, agencyCode)
                : mysqlJdbcTemplate.queryForList(statusSql);

        stats.put("by_status", byStatus);
        return stats;
    }

    public Map<String, Object> reconcileAll(String agencyCode, Integer maxTasks) {
        int limit = maxTasks != null ? maxTasks : 50;
        List<Map<String, Object>> tasks = getPendingTasks(agencyCode, null)
                .stream()
                .limit(limit)
                .collect(Collectors.toList());

        int success = 0;
        int failed = 0;

        for (Map<String, Object> task : tasks) {
            try {
                String taskId = (String) task.get("id");
                Map<String, Object> result = reconcileTask(taskId);
                if ("success".equals(result.get("status"))) {
                    success++;
                } else {
                    failed++;
                }
            } catch (Exception e) {
                log.error("Error reconciling task: {}", e.getMessage());
                failed++;
            }
        }

        Map<String, Object> result = new HashMap<>();
        result.put("success", success);
        result.put("failed", failed);
        result.put("total", tasks.size());

        return result;
    }

    private Map<String, Object> getTask(String taskId) {
        String sql = """
            SELECT t.id, t.ticket_id, t.client_id, t.status, t.created_at, t.attempts,
                   a.client_name
            FROM reconciliation_tasks t
            LEFT JOIN tickets tk ON tk.ticket_number = t.ticket_id
            LEFT JOIN anomalies a ON a.id = tk.anomaly_id
            WHERE t.id = ?
        """;

        try {
            return mysqlJdbcTemplate.queryForMap(sql, taskId);
        } catch (Exception e) {
            return null;
        }
    }

    private List<Map<String, Object>> getCorrections(String ticketId) {
        String sql = """
            SELECT field_name, field_label, old_value, new_value,
                   cbs_value, is_matched, last_checked_at
            FROM corrections
            WHERE ticket_id = ?
        """;

        return mysqlJdbcTemplate.queryForList(sql, ticketId);
    }

    private void updateCorrectionStatus(String ticketId, String field, String cbsValue, boolean isMatched) {
        String sql = """
            UPDATE corrections
            SET cbs_value = ?, is_matched = ?, last_checked_at = NOW()
            WHERE ticket_id = ? AND field_name = ?
        """;

        mysqlJdbcTemplate.update(sql, cbsValue, isMatched, ticketId, field);
    }

    private void updateTaskStatus(String taskId, String status) {
        String sql = """
            UPDATE reconciliation_tasks
            SET status = ?, last_attempt_at = NOW(), attempts = attempts + 1,
                reconciled_at = CASE WHEN ? = 'reconciled' THEN NOW() ELSE reconciled_at END
            WHERE id = ?
        """;

        mysqlJdbcTemplate.update(sql, status, status, taskId);
    }

    private String mapFieldToCBSColumn(String field) {
        Map<String, String> fieldMapping = Map.ofEntries(
                Map.entry("name", "name"),
                Map.entry("firstname", "firstname"),
                Map.entry("address", "address"),
                Map.entry("city", "city"),
                Map.entry("postal_code", "postal_code"),
                Map.entry("phone", "phone"),
                Map.entry("email", "email"),
                Map.entry("birth_date", "birth_date"),
                Map.entry("nationality", "nationality"),
                Map.entry("client_type", "client_type"),
                Map.entry("fatca_status", "fatca_status")
        );

        return fieldMapping.getOrDefault(field, field);
    }

    private boolean compareValues(String expected, String actual) {
        String normalizedExpected = normalizeValue(expected);
        String normalizedActual = normalizeValue(actual);
        return normalizedExpected.equals(normalizedActual);
    }

    private String normalizeValue(String value) {
        if (value == null) return "";
        return value.trim().toLowerCase();
    }

    private String calculateSeverity(String field) {
        List<String> criticalFields = List.of("client_id", "tax_id", "nationality");
        if (criticalFields.contains(field)) return "high";

        List<String> mediumFields = List.of("name", "firstname", "address", "phone", "email");
        if (mediumFields.contains(field)) return "medium";

        return "low";
    }
}
